src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Tds/Tokens/EnvChange/EnvChangeTokenParser.cs

```diff
@@ -0,0 +1,124 @@
+
+                [EnvChangeTokenSubType.ResetConnection] = async (tdsStream, isAsync, ct) =>
+                new ResetConnectionEnvChangeToken(
+                    newValue: await tdsStream.TdsReader.ReadBVarByteAsync(isAsync, ct).ConfigureAwait(false),
+                    oldValue: await tdsStream.TdsReader.ReadBVarByteAsync(isAsync, ct).ConfigureAwait(false)),
+
+                [EnvChangeTokenSubType.Routing] = ParseRoutingEnvChange
+
+            };
+
+        private static async ValueTask<Token> ParseRoutingEnvChange(TdsStream tdsStream, bool isAsync, CancellationToken ct)
+        {
+            ushort length = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+            ByteBuffer buffer = await tdsStream.TdsReader.ReadBufferAsync(length, isAsync, ct).ConfigureAwait(false);
+
+            byte protocol = buffer.ReadUInt8();
+            ushort port = buffer.ReadUInt16LE(1);
+            ushort serverLength = buffer.ReadUInt16LE(3);
+            string server = Encoding.Unicode.GetString(buffer.Slice(5, serverLength * 2).ToArraySegment().Array);
+
+            RoutingInfo newRoutingInfo = new RoutingInfo(protocol, port, server);
+
+            length = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+            RoutingInfo oldRoutingInfo = null;
+
+            if (length > 0)
+            {
+                buffer = await tdsStream.TdsReader.ReadBufferAsync(length, isAsync, ct).ConfigureAwait(false);
+
+                protocol = buffer.ReadUInt8();
+                port = buffer.ReadUInt16LE();
+                serverLength = buffer.ReadUInt16LE(3);
+                server = Encoding.Unicode.GetString(buffer.Slice(5, serverLength * 2).ToArraySegment().Array);
+
+                oldRoutingInfo = new RoutingInfo(protocol, port, server);
+            }
+
+            return new RoutingEnvChangeToken(oldRoutingInfo, newRoutingInfo);
+        }
+
+
+        public override async ValueTask<Token> ParseAsync(TokenType tokenType, TdsStream tdsStream, bool isAsync, CancellationToken ct)
+        {
+            ushort length = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+            byte subType = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+
+            if (!Enum.IsDefined(typeof(EnvChangeTokenSubType), subType))
+            {
+                throw new InvalidOperationException($"Unsupported EnvChange Token type: {subType}");
+            }
+
+            EnvChangeTokenSubType tokenSubType = (EnvChangeTokenSubType)subType;
+
+            if (!subTypeParsers.ContainsKey(tokenSubType))
+            {
+                throw new InvalidOperationException($"Unsupported EnvChange Token type: {tokenSubType}");
+            }
+
+            return await subTypeParsers[tokenSubType](tdsStream, isAsync, ct).ConfigureAwait(false);
```

Comment:
I'm a bit wary of using a dictionary of Funcs, defined with lambdas here. I think we could just as easily have a switch statement for the different sub types that calls the appropriate constructors rather than all the overhead of lambdas and lookups.

Reply:
I agree with Ben, however if you must use a dictionary then use the TryGet method. This way you can avoid 2 lookups at 116 and 121.